### PR TITLE
Find the source container, download and unpack

### PR DESCRIPTION
**Pull request by huiwangredhat**
OLCS-194

- Check the container is a source container

- If it is not the source container, try to find if it has a mapping source container

- Download the source container and unpack it.

### CODE REVIEW

**Comment by rhyw** [comment]
@hanchuntao would you mind give a look at the unpack module in this patch? I guess some of code are based upon legacy repo.

Other that that, I think the latest patch generally LGTM. There could be some minor improvements but we could change along the way. Thanks for the continueous effort on it.

File: openlcs/libs/unpack.py [hunk 1]
```diff
@@ -17,6 +19,27 @@ def get_mime_type(filepath):
...
+        raise RuntimeError(e) from None
+    return directory
+
+
+def run_cli(cmd):
+    try:
+        subprocess.call(cmd, shell=True)
+    except subprocess.CalledProcessError:
+        error = f'Failed to run cmd {cmd}'
+        return '', error
```
Comment (rhyw) [inline openlcs/libs/unpack.py:40]: why would you want to return a message like this, is there a special use case? I didn't see anywhere this `run_cli` is called.

File: openlcs/libs/unpack.py [hunk 2]
```diff
@@ -184,3 +207,56 @@ def unpack_archives(self, src_dir=None):
...
+        """
+        Uncompress tar and tar.gz file according to the file extension.
+        """
+        if f_type == 'tar':
+            cmd = f'tar xvf {src_file} -C {dest_dir}'
+            try:
+                subprocess.call(cmd, shell=True)
+            except subprocess.CalledProcessError:
+                error = f'Failed to uncompress file {src_file}'
+                return '', error
```
Comment (rhyw) [inline openlcs/libs/unpack.py:221]: hmm, if this is where the shortcut `run_cli` resides, please update. also I don't see how the returned tuple is being processed, which is confusing to me.

File: openlcs/libs/unpack.py (hunk 1 above)
Comment (huiwangredhat) [reply openlcs/libs/unpack.py]: Deleted this function that was designed before the function uncompress_tar_file.

File: openlcs/libs/unpack.py (hunk 2 above)
Comment (huiwangredhat) [reply openlcs/libs/unpack.py]: Updated.

File: openlcs/libs/unpack.py [hunk 3]
```diff
@@ -184,3 +199,56 @@ def unpack_archives(self, src_dir=None):
             errors.extend(
                 self.unpack_archives_using_atool(src_dir, top_dir=True))
         return errors
+
+    def uncompress_tar_file(self, src_file=None, dest_dir=None, f_type='tar'):
```
[the remark this answers is not in the document]
Comment (huiwangredhat) [reply openlcs/libs/unpack.py:203]: I checked the [yarn remote source container](https://brewweb.engineering.redhat.com/brew/buildinfo?buildID=2069305), and it has no tgz.  Let's update the code if the tgz is found in the future.

File: openlcs/libs/kojiconnector.py [hunk 4]
```diff
@@ -131,6 +131,34 @@ def get_build(self, build_info):
...
 
+    def get_binary_nvr(self, sc_nvr):
+        """
+        Accept source container nvr that the brew build info is
+        extra.osbs_build.kind: source_container_build
+        Return the mapping binary nvr for the source build
+        """
+        build = self.get_build(sc_nvr)
+        try:
+            if build['extra']['osbs_build']['kind'] ==\
```
[the remark this answers is not in the document]
Comment (huiwangredhat) [reply openlcs/libs/kojiconnector.py]: Updated.

File: openlcs/libs/download.py [hunk 5]
```diff
@@ -84,8 +85,40 @@ def list_packages(self, tag_id, inherited=False):
...
-        return self.koji_connector.list_builds(package_id)
+        # 1.Get the possible source image nvr from binary image nvr
+        nvr = koji.parse_NVR(binary_nvr)
+        name = nvr.get('name')
+        sc_name = name + '-source'
+        n_v_r = (sc_name, nvr.get('version'), nvr.get('release'))
+        sc_nvr = "-".join(n_v_r)
+        # 2.Get the possible source image package name and package id
+        package_name = sc_name
+        package_id = self.koji_connector.get_package_id(package_name)
```
[the remark this answers is not in the document]
Comment (huiwangredhat) [reply openlcs/libs/download.py]: Updated.

File: openlcs/libs/unpack.py [hunk 6]
```diff
@@ -17,6 +19,19 @@ def get_mime_type(filepath):
     return mime_type
 
 
+def create_dir(directory):
```
[the remark this answers is not in the document]
Comment (huiwangredhat) [reply openlcs/libs/unpack.py]: Updated.

File: openlcs/libs/unpack.py [hunk 7]
```diff
@@ -184,3 +199,56 @@ def unpack_archives(self, src_dir=None):
...
+        else:
+            err_msg = "The file {src_file} is unsupported to uncompress."
+            raise ValueError(err_msg) from None
+        os.remove(src_file)
+
+    def unpack_source_container_image(self, src_file=None, dest_dir=None):
+        """
+        Unpack source container image.
+        """
+        # Unpack the top level of source cotainer image
```
[the remark this answers is not in the document]
Comment (huiwangredhat) [reply openlcs/libs/unpack.py]: updated.

File: openlcs/libs/unpack.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (huiwangredhat) [reply openlcs/libs/unpack.py:191]: Deleted the src_file default value None.

File: openlcsd/flow/tasks.py [hunk 8]
```diff
@@ -149,6 +146,10 @@ def get_source_container_build(context, engine):
         build_type = koji_build.get_build_type(sc_build)
         context['build'] = sc_build
         context['build_type'] = build_type
+        context['container_type'] = 'source_container_build'
```
[the remark this answers is not in the document]
Comment (huiwangredhat) [reply openlcsd/flow/tasks.py]: Updated.

File: openlcsd/flow/tasks.py [hunk 9]
```diff
@@ -447,6 +448,37 @@ def unpack_source(context, engine):
...
+    try:
+        misc_dir, srpm_dir, rc_dir = \
+                ua.unpack_source_container_image(src_file, src_dest_dir)
+    except RuntimeError as e:
+        err_msg = "Failed to decompress file %s: %s" % (src_file, e)
+        raise RuntimeError(err_msg) from None
+    engine.logger.info('Finished unpack the source archives.')
+    context['misc_dir'] = misc_dir
+    context['srpm_dir'] = srpm_dir
+    context['rc_dir'] = rc_dir
```
[the remark this answers is not in the document]
Comment (huiwangredhat) [reply openlcsd/flow/tasks.py]: Updated.

File: openlcsd/flow/tasks.py [hunk 10]
```diff
@@ -447,6 +448,37 @@ def unpack_source(context, engine):
...
+    src_dest_dir = context.get('src_dest_dir')
+    src_file = os.path.join(src_dest_dir, file)
+    # Unpack the source container image
+    ua = UnpackArchive(config=config)
+    engine.logger.info('Start to unpack source container image...')
+    try:
+        misc_dir, srpm_dir, rc_dir = \
+                ua.unpack_source_container_image(src_file, src_dest_dir)
+    except RuntimeError as e:
+        err_msg = "Failed to decompress file %s: %s" % (src_file, e)
```
[the remark this answers is not in the document]
Comment (huiwangredhat) [reply openlcsd/flow/tasks.py:467]: Updated.

File: openlcsd/flow/tasks.py [hunk 11]
```diff
@@ -447,6 +448,37 @@ def unpack_source(context, engine):
...
+        raise RuntimeError(err_msg) from None
+    engine.logger.info('Finished unpack the source archives.')
```
[the remark this answers is not in the document]
Comment (huiwangredhat) [reply openlcsd/flow/tasks.py]: Updated.

File: openlcsd/flow/tasks.py [hunk 12]
```diff
@@ -447,6 +448,37 @@ def unpack_source(context, engine):
...
 
+def unpack_container_source_archive(context, engine):
+    """
+    Uncompress sources of given build/archive.
+
+    @requires: 'src_dest_dir', destination directory.
+    @requires: 'src_file'.
+    """
+    prepare_dest_dir(context, engine)
+    extract_source(context, engine)
```
[the remark this answers is not in the document]
Comment (huiwangredhat) [reply openlcsd/flow/tasks.py:452]: In the task.py, line 356 and line line 403.

File: openlcsd/flow/tasks.py [hunk 13]
```diff
@@ -704,10 +732,16 @@ def fork_component_imports(context, engine):
...
+            IF(
+                lambda o, e: o.get('container_type') == 'source_container_build', # noqa
+                [
+                    # Download the source image from Brew
+                    download_source_image,
+                    # Unpack the source image
+                    unpack_container_source_archive,
+                ],
+            ),
             # Get the source path for each component from contaner source
```
[the remark this answers is not in the document]
Comment (huiwangredhat) [reply openlcsd/flow/tasks.py]: Updated.

**Comment by yulwang0623** [comment]
https://issues.redhat.com/browse/OLCS-240

**Comment by hanchuntao** [comment]
LGTM

**Pull request merged**